core/src/main/java/org/apache/iceberg/ReplaceManifests.java

```diff
@@ -0,0 +1,221 @@
+  private Function<ManifestFile, Boolean> filterFunc;
+
+  ReplaceManifests(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.spec = ops.current().spec();
+    this.manifestTargetSizeBytes =
+      ops.current().propertyAsLong(MANIFEST_TARGET_SIZE_BYTES, MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+  }
+
+  @Override
+  protected String operation() {
+    return DataOperations.REPLACE;
+  }
+
+  @Override
+  public RewriteManifests set(String property, String value) {
+    summaryBuilder.set(property, value);
+    return this;
+  }
+
+  @Override
+  protected Map<String, String> summary() {
+    return summaryBuilder.build();
+  }
+
+  @Override
+  public ReplaceManifests clusterBy(Function<DataFile, Object> func) {
+    this.clusterByFunc = func;
+    return this;
+  }
+
+  @Override
+  public ReplaceManifests filter(Function<ManifestFile, Boolean> func) {
+    this.filterFunc = func;
+    return this;
+  }
+
+  @Override
+  public List<ManifestFile> apply(TableMetadata base) {
+    Preconditions.checkNotNull(clusterByFunc, "clusterBy function cannot be null");
+
+    List<ManifestFile> existingManifests = Collections.synchronizedList(new ArrayList<>());
+
+    try {
+      Tasks.foreach(base.currentSnapshot().manifests())
+          .executeWith(ThreadPools.getWorkerPool())
+          .run(manifest -> {
+            if (filterFunc != null && !filterFunc.apply(manifest)) {
+              existingManifests.add(manifest);
+            } else {
+              long entryNum = manifest.addedFilesCount() + manifest.existingFilesCount() + manifest.deletedFilesCount();
+              long avgEntryLen = manifest.length() / entryNum;
+
+              try (ManifestReader reader =
+                     ManifestReader.read(ops.io().newInputFile(manifest.path()))) {
+                FilteredManifest filteredManifest = reader.select(Lists.newArrayList("*"));
+                filteredManifest.iterator().forEachRemaining(
+                    file -> appendFile(file, avgEntryLen, clusterByFunc.apply(file))
+                );
+
+              } catch (IOException x) {
+                throw new RuntimeIOException(x);
+              }
+            }
+          });
+    } finally {
+      Tasks.foreach(writers.values()).executeWith(ThreadPools.getWorkerPool()).run(writer -> writer.close());
+    }
+
+    newManifests.addAll(
+        writers.values()
+          .stream()
+          .map(WriterWrapper::toManifestFile)
+          .collect(Collectors.toList())
+    );
+
+    // put new manifests at the beginning
+    List<ManifestFile> apply = new ArrayList<>();
+    apply.addAll(newManifests);
+    apply.addAll(existingManifests);
+
+    return apply;
+  }
+
+  private void appendFile(DataFile file, long avgEntryLen, Object key) {
+    Preconditions.checkNotNull(file, "Data file cannot be null");
+    Preconditions.checkNotNull(key, "Key cannot be null");
+
+    synchronized (summaryBuilder) {
+      summaryBuilder.addedFile(spec, file);
```

Comment:
@bryanck Looks like this is potentially pulling all the partitions into memory during this process.  While this may work for smaller cases, I doubt this will scale given the current summary builder.  We could not write these at all since this is a replace operation and the partition summaries won't be used for any current use case.  I assume we would fallback elsewhere if the summaries don't exist.

Reply:
This may be an issue in general, i.e. for updates that span many partitions. Perhaps I could add an option not to track number of partitions modified to summary builder, or stop tracking if the number reaches some threshold. Any thoughts?

Reply:
Looking through the code it appears leaving the summary empty/zero won't break anything. I think it makes sense to keep it empty given this is a replace. I'll make that change...

Reply:
I agree, this is okay because it isn't actually calling addFile or deleteFile.